### ISSUE TITLE
p4test fails for ArrayIndex expressions with non-constant  indices

### ISSUE THREAD

**Issue by usha1830** [p4lang/p4c#3374]
p4test fails when ArrayIndex expressions with non-constant different indices occur on both sides of assignment and when such an expression occurs as Switch expression. 

The following statements fail to compile with p4test. Is this expected?

```
1) hdrs.vlan_tag[meta.depth].vid = hdrs.vlan_tag[meta.depth-1].vid;//  line 97 in attached program

// replacing one side array index with constant works fine

2) switch (hdrs.vlan_tag[meta.depth].vid)   // line 129 in attached program

 // replacing meta.depth with constant works fine

```

The error is :
_pna-example-varIndex.p4(98): [--Werror=type-error] error: AssignmentStatement
        hdrs.vlan_tag[meta.depth].vid = hdrs.vlan_tag[meta.depth-1].vid;
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ---- Actual error:
  Cannot unify type 'bit<16>' with type 'bit<12>'
  ---- Originating from:
  pna-example-varIndex.p4(94): Source expression 'hsVar_1' produces a result of type 'bit<16>' which cannot be assigned to a left-value with type 'bit<12>'
          meta.ethType = hdrs.vlan_tag[meta.depth - 1].ether_type;
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pna-example-varIndex.p4(130): [--Werror=type-error] error: AssignmentStatement
        switch (hdrs.vlan_tag[meta.depth].vid) {
                ^^^^^^^^^^^^^^^^^^^^^^^^^
  ---- Actual error:
  Cannot unify type 'Unknown type' with type 'bit<12>'
  ---- Originating from:
  pna-example-varIndex.p4(130): Source expression 'hsVar' produces a result of type 'Unknown type' which cannot be assigned to a left-value with type 'bit<12>'
          switch (hdrs.vlan_tag[meta.depth].vid) {_

Example P4 program is attached.
[hsindex.p4.txt](https://github.com/p4lang/p4c/files/8805336/hsindex.p4.txt)

**Comment by jafingerhut** [comment]
@usha1830 I tried running this command `p4test hsindex.p4` with your attached P4 program, using this version of the p4c source code, unmodified:
```
$ git log -n 1 | head -n 3
commit f4ca7ff4f5c5f2f701ee110447915fca629f9cea
Author: Mihai Budiu <mbudiu@vmware.com>
Date:   Tue May 31 08:43:23 2022 +0200
```
I got no error messages at all.

Can you try with a version of p4test built from that source code to see if you still get errors?

**Comment by usha1830** [comment]
> @usha1830 I tried running this command `p4test hsindex.p4` with your attached P4 program, using this version of the p4c source code, unmodified:
> 
> ```
> $ git log -n 1 | head -n 3
> commit f4ca7ff4f5c5f2f701ee110447915fca629f9cea
> Author: Mihai Budiu <mbudiu@vmware.com>
> Date:   Tue May 31 08:43:23 2022 +0200
> ```
> 
> I got no error messages at all.
> 
> Can you try with a version of p4test built from that source code to see if you still get errors?

@jafingerhut My bad!! I commented the lines having this error. 
Attached the modified test now.
[hsindex.p4.txt](https://github.com/p4lang/p4c/files/8810683/hsindex.p4.txt)

**Comment by usha1830** [comment]
1`) hdrs.vlan_tag[meta.depth].vid = hdrs.vlan_tag[meta.depth-1].vid;//  line 97 in attached program`
The above statement compiles fine when other statements in the action before this statement are removed.

I have added a few tests in this PR https://github.com/p4lang/p4c/pull/3375 for testing dynamic index access for header stack elements
One test which used it within switch is still failing with p4test.

**Comment by jafingerhut** [comment]
I don't have any more details on the root cause of why these error messages are produced by the `p4test` command, but from an examination of the P4 program, the error messages definitely look incorrect, and so this does appear to be a bug in p4test somewhere.

**Comment by usha1830** [comment]
@jafingerhut Thanks for checking. I too don't have more details as of now. I will take a look at it and try to fix it.
@jafingerhut @mbudiu-vmw Can we keep this issue open and mark it as a bug?

**Comment by mbudiu-vmw** [comment]
i am vacation, but today it's raining so I will try to take a look.

**Comment by mbudiu-vmw** [comment]
BTW: this code was written by @VolodymyrPeschanenkoIntel.
The bug seems to be that he reuses the same uninitialized variable to represent an illegal index even if the index is used in multiple places. The code he generates looks like this:
```
    bit<16> hsVar_1;
    @name("MainControlImpl.execute") action execute_1() {
        hsiVar_4 = meta.depth + 2w3;
        if (hsiVar_4 == 2w0) {
            meta.ethType = hdrs.vlan_tag[2w0].ether_type;
        } else if (hsiVar_4 == 2w1) {
            meta.ethType = hdrs.vlan_tag[2w1].ether_type;
        } else if (hsiVar_4 >= 2w1) {
            meta.ethType = hsVar_1; // out of bounds index
        }
       // The code above is fine, removing some code and:
        hsiVar_5 = meta.depth;
        if (hsiVar_5 == 2w0) {
            hsiVar_4 = meta.depth + 2w3;
            if (hsiVar_4 == 2w0) {
                hdrs.vlan_tag[2w0].vid = hdrs.vlan_tag[2w0].vid;
            } else if (hsiVar_4 == 2w1) {
                hdrs.vlan_tag[2w0].vid = hdrs.vlan_tag[2w1].vid;
            } else if (hsiVar_4 >= 2w1) {
                hdrs.vlan_tag[2w0].vid = hsVar_1;
            }
        }
    }
```

Notice how hsVar_1 is used to initialize both .ethType and .vid, which have different types.
Different variables should be used for this purpose.

**Comment by mbudiu-vmw** [comment]
The way I would write the array index "constantification" pass is as follows:
- convert each expression used as an array index which is not constant into a temporary variable.
- convert each array ArrayIndex expression with a non-constant index (except the ones that are left values) into a temporary variable as well.
  This would ensure that there aren't two ArrayIndex subexpressions in the same expression, and that such expressions do not appear in table keys, switch statements, etc.
- finally, convert each ArrayIndex expression (on the RHS or LHS of an assignment statement) into a sequence of if statements.
  Each statement should have at most one such expression, and all such expressions should be in an AssignmentStatement, the single expression on the LHS or RHS of a statement.
@usha1830 do you want to try to rewrite this pass in such a way? I haven't read your original submission, but perhaps it's not too far.

**Comment by hesingh** [comment]
I added support to the bmv2 backend. See

https://github.com/p4lang/p4c/pull/2118

**Comment by mbudiu-vmw** [comment]
@VolodymyrPeschanenkoIntel there are actually several different bugs; the type mismatch bug can be fixed easily by just using toString() instead of type_name(). The second one is harder to fix, it is about a field with unknown type. In general, you should not use expression->type, that is not up-to-date, you need to lookup types in the typemap. However, some expressions are generated by other passes you invoke, and thus their types will not be in the typemap. Also, the pass relies on the typemap being correct, but in general if you want this to hold you have to invoke typeChecking explicitly in your pass.

So I think the pass will need significant refactoring.

**Comment by usha1830** [comment]
> The way I would write the array index "constantification" pass is as follows:
> 
> * convert each expression used as an array index which is not constant into a temporary variable.
> * convert each array ArrayIndex expression with a non-constant index (except the ones that are left values) into a temporary variable as well.
>   This would ensure that there aren't two ArrayIndex subexpressions in the same expression, and that such expressions do not appear in table keys, switch statements, etc.
> * finally, convert each ArrayIndex expression (on the RHS or LHS of an assignment statement) into a sequence of if statements.
>   Each statement should have at most one such expression, and all such expressions should be in an AssignmentStatement, the single expression on the LHS or RHS of a statement.
>   @usha1830 do you want to try to rewrite this pass in such a way? I haven't read your original submission, but perhaps it's not too far.

Implementation in my original[ PR](https://github.com/p4lang/p4c/pull/3371) is close to this design. except that I have not simplified the Assignment statements having Arrayindex expressions on both sides. Those are directly handled while converting to if statements.

For simplifying the expressions within various statements, existing pass in DPDK "StatementUnroll" is updated to handle the ArrayIndex expression.

**Comment by usha1830** [comment]
@VolodymyrPeschanenkoIntel  Another observation in using the current HSSimplifier pass is, when ArrayIndex is used in switch expression, the output is as follows:

```
P4: 
        switch (hdrs.vlan_tag[meta.depth].vid) {
            12w1: { stub.apply();}
            12w2: { stub1.apply();}
        }
```

Output after HSSimplifier:
  ```
      if (hsiVar == 2w0) {
            switch (hdrs.vlan_tag[2w0].vid) {
                12w1: {
                    {
                        key_0 = hdrs.vlan_tag[2w0].vid;
                        stub_0.apply();
                    }
                }
                12w2: {
                    stub1_0.apply();
                }
            }
        } else if (hsiVar == 2w1) {
            switch (hdrs.vlan_tag[2w1].vid) {
                12w1: {
                    {
                        key_0 = hdrs.vlan_tag[2w1].vid;
                        stub_0.apply();
                    }
                }
                12w2: {
                    stub1_0.apply();
                }
            }
        }
```
If we take the ArrayIndex expression in temporary as suggested, it will be optimal.

**Comment by mbudiu-vmw** [comment]
I think that this issue is actually fixed by #3382, so I will close it. Feel free to reopen if you think this is incorrect.
